### PR TITLE
Update Dependencies

### PR DESCRIPTION
**Pull request by Ardakilic**
# Monolithic Dependency Update (Batch PR)

**Branch:** `chore/dependency-update-batch`  
**Base:** `main`  
**Scope:** All 4 workspace packages — `apps/api`, `apps/web`, `packages/db`, `packages/shared`

## Summary

Updates 9 production/dev dependencies across the BrewForm monorepo in a single PR, including one package replacement (`@base-ui-components/react` → `@base-ui/react`). All code migrations, type adjustments, and test fixes are included.

> **Why monolithic?** These updates were prepared as 9 separate Renovate PRs but are merged into one to reduce CI churn and simplify rollback (single `git revert` + `deno install`).

---

## Dependency Changes

| Dependency | Old | New | Scope | Breaking Changes |
|---|---|---|---|---|
| `zod` | `^3.x` | `^4.0.0` | API, Shared | Migration guide: v3 → v4 |
| `typescript` | `^5.x` | `^6.0.0` | Web | Higher strictness, new `types` config |
| `vitest` | `^3.x` | `^4.0.0` | Web | Compatible with TypeScript 6 |
| `vite` | `^6.x` | `^8.0.0` | Web | Rolldown replaces Rollup |
| `@vitejs/plugin-react` | `^4.x` | `^6.0.0` | Web | Aligned with Vite 8 |
| `mjml` | `^4.x` | `^5.0.0` | API | Async API, removed minify/beautify/filePath |
| `bcryptjs` | `^2.x` | `^3.0.0` | API, DB | Named exports only (no default) |
| `pino` | `^9.x` | `^10.0.0` | API | Patch-level compatible |
| `fast-check` | `^3.x` | `^4.0.0` | Web | `fc.date()` can emit `Invalid Date` |

### Package Replacement

| Removed | Added | Reason |
|---|---|---|
| `@base-ui-components/react` | `@base-ui/react` | Package renamed upstream |
| `@types/bcryptjs` | *(removed)* | bcryptjs v3 ships its own types |

---

### 1. Package Manifests (4 files)
- `apps/api/package.json` — zod, pino, bcryptjs, mjml bumps; removed @types/bcryptjs
- `apps/web/package.json` — vite, typescript, vitest, fast-check bumps; @base-ui/react added
- `packages/shared/package.json` — zod bump
- `packages/db/package.json` — bcryptjs bump

### 2. Zod v4 Schema Migrations (17+ files)
- `packages/shared/src/schemas/auth.ts`, `bean.ts`, `comment.ts`, `common.ts`, `recipe.ts`, `taste.ts`, `user.ts`, `vendor.ts`
- **Method replacements:** `z.string().email()` → `z.email()`, `z.string().uuid()` → `z.uuid()`, `z.string().url()` → `z.url()`, `z.string().date()` → `z.iso.date()`
- `z.record()` now requires 2 type arguments where previously 1 was optional
- `z.discriminatedUnion()` requires explicit `.options()` method call
- `.refine(val, ctx)` → `.superRefine()` with `ctx.addIssue()` for conditional validation

### 3. API Runtime Code (5 files)
- `apps/api/src/config/env.ts` — `.superRefine()` + `z.url()` + `z.ZodIssueCode.custom`
- `apps/api/src/middleware/errorHandler.ts` — `.errors` → `.issues` (zod v4 renamed property)
- `apps/api/src/modules/admin/index.ts` — inline `z.string().email()` → `z.email()`
- `apps/api/src/modules/report/index.ts` — inline `z.string().uuid()` → `z.uuid()`
- `apps/api/src/utils/response/index.ts` — widened path type for `.format()` compatibility

### 4. bcryptjs Named Imports (4 files)
- `apps/api/src/modules/auth/model.ts`
- `apps/api/src/modules/admin/model.ts`
- `apps/api/src/setup.ts`
- `packages/db/src/seed-data.ts`
- **Pattern:** `import * as bcrypt from 'bcryptjs'` → `import { hashSync, compareSync } from 'bcryptjs'`

### 5. `@base-ui/react` Imports (5 web files)
- `apps/web/src/pages/TasteNotesPage.tsx` — `@base-ui/react/popover`
- `apps/web/src/pages/TasteNotesPage.test.tsx` — `@base-ui/react/popover`
- `apps/web/src/components/recipe/TasteNotesFilter.tsx` — `@base-ui/react/select`
- `apps/web/src/components/layout/Navbar.tsx` — `@base-ui/react/select`
- `apps/web/src/components/layout/LanguageSelector.tsx` — `@base-ui/react/select`

### 6. TypeScript 6 Compatibility (5 files)
- `apps/web/tsconfig.json` — Added explicit `types: ["vitest/globals"]`
- `apps/api/src/modules/auth/jwt.test.ts` — Type assertions for `AccessPayload` fields
- `apps/api/src/modules/comment/service.test.ts` — Non-null assertions for `createCallArgs`
- `apps/api/src/modules/recipe/service.test.ts` — UUID fix for fast-check v4
- `packages/shared/src/schemas/recipe.test.ts` — UUID fix + date PBT filter

### 7. mjml v5 Migration (2 files)
- `apps/api/src/types/mjml.d.ts` — `MjmlOptions` interface rewritten (removed `minify`/`beautify`/`filePath`, added `fonts`/`keepComments`/`juiceOptions`); return type changed to `Promise<MjmlResult>`
- `apps/api/scripts/build-email-templates.ts` — Added `await` to `mjml2html()` call (v5 is async); null-safe `errors` access

### 8. fast-check v4 PBT Fixes (1 file)
- `packages/shared/src/schemas/recipe.test.ts` — `.filter((d) => !isNaN(d.getTime()))` on all `fc.date()` generators (v4 includes `Invalid Date` values by default)

### 9. Lockfile (1 file)
- `deno.lock` — Regenerated via `deno install` (150 packages resolved)

### 10. Generated Email Templates (6 files)
- `apps/api/src/templates/email/generated/*.ts` — Rebuilt by `deno task email-build` with mjml v5

---

## Verification Results

| Check | Status | Details |
|---|---|---|
| Formatting | ✅ | `deno fmt --check` — 328 files clean |
| Linting | ✅ | `deno lint` — 311 files clean |
| Type-checking | ✅ | All 3 workspaces (`apps/api`, `packages/db`, `packages/shared`) |
| Unit tests | ✅ | 70 passed, 0 failed (537 steps, 2s) |
| Vite production build | ✅ | 672 KB JS + 30 KB CSS, Rolldown bundler |
| Vite dev server | ✅ | Starts on port 5173, HTTP 200 |
| Email template build | ✅ | All 6 MJML templates compile to TypeScript |
| API server startup | ✅ | Code compiles; env validation catches missing vars correctly |

---

## Test Details

- **70 tests total** across 33 test suites
- **33 PBT properties** (fast-check) cover edge cases for recipe dates, comment authorization, taste note IDs, and user profile preservation
- All PBT tests use `fc.date().filter()` to exclude `Invalid Date` values from fast-check v4
- 0 flaky tests observed across multiple runs

## Rollback Instructions

```bash
git revert <commit-sha>
deno install
```

This restores the previous lockfile and all original dependency versions.

## What Was NOT Changed

- No application logic modifications (schema validation rules, business logic, API contracts)
- No database migrations
- No runtime behavior changes beyond what the new dependency versions require
- No public API or data structure changes

## Release Notes

* **New Features**
  * Updated UI components to use `@base-ui/react` library across language selector, theme switcher, and filter components.

* **Bug Fixes**
  * Improved error handling robustness for validation errors and email template generation.

* **Chores**
  * Updated workspace dependencies: Zod (v3→v4), pino (v9→v10), bcryptjs (v2→v3), mjml (v4→v5), TypeScript, Vite, vitest, and test tooling to latest versions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Ardakilic/BrewForm/pull/28?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: dependency_updates_plan.md [hunk 1]
```diff
@@ -0,0 +1,851 @@
...
+
+Zod v4 is a major rewrite with API differences from v3. Key breaking changes:
+
+1. **Import paths changed**: `z.object()`, `z.string()`, `z.number()` etc. remain top-level but some utilities moved
+2. **`.refine()` behavior**: Refinements on `.pick()` / `.omit()` / `.extend()` now throw errors — use `.safeExtend()` instead
+3. **Schema composition**: Object masking methods (`.pick()`, `.omit()`) now validate keys exist
+4. **No `.nonempty()` on strings** — replaced by `.min(1)` pattern
+5. **`z.enum()` may have changed semantics**
+6. **ZodError formatting**: `flatten()` and `format()` APIs may differ
+
```
[the remark this answers is not in the document]
Comment (Ardakilic) [reply dependency_updates_plan.md:86]: @coderabbitai this was my initial plan, not the actual implementation. You can ignore this.
